src/wg/format/bmp.d

```diff
@@ -0,0 +1,294 @@
+                return 4;
+            case BMPInfoV5.sizeof:
+                return 5;
+            default:
+                return 0;
+        }
+    }
+
+    const(Info)* info(Info)() const
+    {
+        return cast(const(Info)*)(header() + 1);
+    }
+
+    const(void)[] iccData() const
+    {
+        if (bmpVersion() < 5)
+            return null;
+        const(BMPInfoV5)* i = info!BMPInfoV5();
+        return (cast(void*)i + i.profileData)[0 .. i.profileSize];
+    }
+
+    inout(ImageBuffer) getImage() inout
+    {
+        ImageBuffer img;
+        CommonMetadata md;
+
+        string cs = "sRGB";
+        string format;
+        ushort bits;
+        uint width, height;
+        bool topDown = false;
+
+        // TODO: no paletted images...
+
+        final switch(bmpVersion())
+        {
+            case 2:
+                const(BMPInfoV2)* i = info!BMPInfoV2();
+
+                assert(i.planes == 1);
+                bits = i.bitCount;
+
+                width = i.width;
+                height = i.height;
+                break;
+            case 5:
+                const(BMPInfoV5)* i = info!BMPInfoV5();
+
+                // what is 'intent'?
+
+                goto case 4;
+            case 4:
+                const(BMPInfoV4)* i = info!BMPInfoV4();
+
+                // parse the channel masks...
+                if (i.compression == Compression.BI_BITFIELDS)
+                {
+//                    i.redMask;
+//                    i.greenMask;
+//                    i.blueMask;
+//                    i.alphaMask;
+                }
+
+                goto case 3;
+            case 3:
+                const(BMPInfoV3)* i = info!BMPInfoV3();
+
+                assert(i.planes == 1);
+                bits = i.bitCount;
+
+                assert(i.width >= 0);
+                width = i.width;
+
+                if (i.height < 0)
+                {
+                    topDown = true;
+                    height = -i.height;
+                }
+                else
+                    height = i.height;
+
+                assert(i.clrUsed == 0);
+
+                // check 'compression'
+
+                md.horizDpi = i.xPelsPerMeter / 39.37007874;
+                if (i.xPelsPerMeter && i.yPelsPerMeter)
+                    md.pixelAspect = cast(float)i.xPelsPerMeter / cast(float)i.yPelsPerMeter;
+                break;
+        }
+
+        if (!format) switch(bits)
+        {
+            case 16: format = "bgrx_5_5_5_1"; break;
+            case 24: format = "bgr";          break;
+            case 32: format = "bgrx";         break;
+            default:
+                // 0, 1, 4, 8
+                assert(0);
+        }
+
+        return inout(ImageBuffer)(
+            width, height,
+            (((bits * width) / 8) + 3) & ~3,
+            1, 1,
+            cast(ubyte)bits,
+            0,
+            data.ptr + header().offBits,
+            format.ptr,
+            null
+        );
```

Comment:
This is so unreadable without named parameters :(. Do we have some alternative to make this more readable?

Reply:
if a parameter is unreadable, you can always have a named variable in the line before.

Reply:
Yeah D's `const` is a problem here; I couldn't just declare a mutable struct locally and populate it, and then return it... I had to initialise the const struct in-place.
I hate that code, I'll definitely rewrite most of it; but I just wanted something quick and dirty to get real pictures into the system for experiments.